aplib.net-demo/Assets/Scripts/Player/Movement.cs

```diff
@@ -0,0 +1,147 @@
+
+    /// <summary>
+    /// Initialize the player's rigidbody and collider, and freeze the player's rotation.
+    /// </summary>
+    void Start() { 
+        rb = GetComponent<Rigidbody>();
+        controller = GetComponent<CapsuleCollider>();
+        playerHeight = controller.height;
+        rb.freezeRotation = true;
+        gravity = Physics.gravity.y;
+    }
+
+    /// <summary>
+    /// Main update loop where we do a ground check, move the player, and handle jumping.
+    /// Updates once per physics update.
+    /// </summary>
+    void FixedUpdate() {
+        // Ground check by checking a sphere at the bottom of the player, more consistent than ray
+        Vector3 bottomPoint = transform.TransformPoint(controller.center - 0.5f * playerHeight * Vector3.up);
+        isGrounded = Physics.CheckSphere(bottomPoint, 0.15f, groundMask);
+
+        MovePlayer();
+        HandleJump();
+    }   
+    
+    /// <summary>
+    /// Move the player based on the input and the ground check using AddForce.
+    /// Also applies custom gravity and limits the player's velocity.
+    /// </summary>
+    void MovePlayer() {
+        horizontalVelocity = transform.right * horizontalInput.x + transform.forward * horizontalInput.y;
+        bool isOnSlope = OnWalkableSlope(out Vector3 directionOnSlope, out RaycastHit downHit);
+        rb.useGravity = !isOnSlope;
+
+        if (isGrounded) {
+            rb.drag = groundDrag; 
+            Debug.Log("Grounded");
+            if (isOnSlope) {
+                rb.AddForce(maxSpeed * acceleration * Time.fixedDeltaTime * 1.5f * directionOnSlope);
+                // Apply gravity but perpendicular to the slope, to prevent sliding
+                rb.velocity += (actualGravityScale - 1) * gravity * Time.fixedDeltaTime * downHit.normal;
+            }
+            else {
+                rb.AddForce(maxSpeed * acceleration * Time.fixedDeltaTime * horizontalVelocity.normalized);
+            }   
+        }
+        else {
+            rb.drag = 0;  
+            rb.AddForce(maxSpeed * acceleration * Time.fixedDeltaTime * airMovementScale * horizontalVelocity.normalized); 
+            // Custom gravity for player while falling
+            actualGravityScale = rb.velocity.y < 0 ? fallingGravityScale : normalGravityScale;
+            rb.velocity += (actualGravityScale - 1) * gravity * Time.fixedDeltaTime * Vector3.up;
+        }
+
+        LimitVelocity(isOnSlope);
+    }
+
+    /// <summary>
+    ///  Limit the player's velocity to the max speed.
+    ///  Also accounts for vertical velocity when on a slope.
+    /// </summary>
+    /// <param name="isOnWalkableSlope"></param>
+    void LimitVelocity(bool isOnWalkableSlope){
+        if (isOnWalkableSlope) {   
+            if (rb.velocity.magnitude > maxSpeed) {
+                rb.velocity = rb.velocity.normalized * maxSpeed;
+            }
+        } else {
+            Vector3 rbHorizontalVelocity = new(rb.velocity.x, 0, rb.velocity.z);
```

Comment:
Naming suggestion: `rigidbodyHorizontalVelocity` instead of `rbHorizontalVelocity`.